addons/frag/functions/fnc_doFrag.sqf

```diff
@@ -0,0 +1,61 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Jaynus, NouberNou, Lambda.Tiger
+ * This function handles creating both random and targeted fragments as well as handling some of the performance optimizations.
+ *
+ * Arguments:
+ * 0: Position (posASL) of projectile <ARRAY>
+ * 1: Velocity of projectile <ARRAY>
+ * 2: Projectile CfgAmmo classname <STRING>
+ * 3: getShotParents of projectile at EH <ARRAY>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [getPosASL _projectile, velocity _projectile, typeOf _projectile, getShotParents _projectile] call ace_frag_fnc_doFrag;
+ *
+ * Public: No
+ */
+
+TRACE_1("begin doFrag",_this);
+params ["_posASL", "_velocity", "_ammo", "_shotParents"];
+
+// Don't let a single object cause all fragmentation events
+_shotParents params ["_shotParentVic"];
+if (_shotParentVic getVariable [QGVAR(obj_nextFragTime), -1] > CBA_missionTime) exitWith {
+    TRACE_1("vehicleTimeExit",_shotParentVic);
+};
+_shotParentVic setVariable [QGVAR(obj_nextFragTime), CBA_missionTime + ACE_FRAG_HOLDOFF_VEHICLE];
+
+// Check normal round timeout and adjust _max frags
+private _timeSinceLastFrag = CBA_missionTime - GVAR(lastFragTime);
+if (_timeSinceLastFrag < ACE_FRAG_HOLDOFF || {_posASL isEqualTo [0, 0, 0]} || {_ammo isEqualTo ""}) exitWith {
+    TRACE_3("timeExit",_timeSinceLastFrag,CBA_missionTime,GVAR(lastFragTime));
+};
+private _maxFragCount = round linearConversion [ACE_FRAG_COUNT_MIN_TIME, ACE_FRAG_COUNT_MAX_TIME, _timeSinceLastFrag, ACE_FRAG_COUNT_MIN, ACE_FRAG_COUNT_MAX, true];
+TRACE_3("willFrag",_timeSinceLastFrag,CBA_missionTime,_maxFragCount);
+
+private _ammoArr = [_ammo] call FUNC(getFragInfo);
+_ammoArr params ["_fragRange", "_fragVel", "_fragTypes", "_modFragCount"];
+// For low frag rounds limit the # of frags created
+if (_modFragCount < ACE_FRAG_LOW_FRAG_MOD_COUNT) then {
+    _maxFragCount = _modFragCount * ACE_FRAG_LOW_FRAG_COEFF;
+    GVAR(lastFragTime) = CBA_missionTime - ACE_FRAG_LOW_FRAG_HOLDOFF_REDUCTION;
+} else {
+    GVAR(lastFragTime) = CBA_missionTime;
+};
+// Offset for ground clearance
+private _heightATL = (ASLToATL _posASL)#2;
+if (_heightATL < ACE_FRAG_MIN_GROUND_OFFSET) then {
+    _posASL = _posASL vectorAdd [0, 0, ACE_FRAG_MIN_GROUND_OFFSET - (0 min _heightATL)];
+};
```

Comment:
Ok, so this group of statements is based on [this line](https://github.com/acemod/ACE3/blob/64538f2ad00a5a400b7823aef01e5e4c4a23044c/addons/frag/functions/fnc_frago.sqf#L80) in the frago code. I believe the original intention was to avoid fragments being spawned underground, but using the BI explode EH I haven't run into the issue. I think it made sense in some respects, but in some rough terrain or near low walls you may be hit by fragments when you might have had (very low) cover from the explosion. I think it's fine to get rid of it and change the downstream code accordingly, but does anyone have opinions as why to leave it?

Reply:
Changed here: https://github.com/acemod/ACE3/pull/9728/commits/deecaf9249003bdf784f68219877330c48c501dc. Also, I would like to note that the original/current version does this `vectorAdd` needlessly as it never saves the value returned from the operation. So if this PR doesn't end up getting merged for some reason, it's still worth removing those lines.